src/Common/src/Interop/Unix/libgss/Interop.GssApi.cs

```diff
@@ -0,0 +1,162 @@
+            Debug.Assert((count > 0) && (count <= (buffer.Length - offset)), "Invalid input count passed to Encrypt");
+
+            using (SafeGssBufferHandle outputToken = new SafeGssBufferHandle())
+            {
+                libgssapi.Status minorStatus;
+                libgssapi.Status status = libgssapi.GssWrap(out minorStatus, context, encrypt, buffer, offset, count, outputToken);
+                if (status != libgssapi.Status.GSS_S_COMPLETE)
+                {
+                    throw libgssapi.GssApiException.Create(SR.net_context_wrap_failed, status, minorStatus);
+                }
+
+                outputBuffer = new byte[outputToken.Length]; // Always return non-null
+                if (outputToken.Length > 0)
+                {
+                    Marshal.Copy(outputToken.Value, outputBuffer, 0, outputToken.Length);
+                }
+
+                return outputBuffer.Length;
+            }
+        }
+
+        internal static int Decrypt(
+            SafeGssContextHandle context,
+            byte[] buffer,
+            int offset,
+            int count)
+        {
+            Debug.Assert((buffer != null) && (buffer.Length > 0), "Invalid input buffer passed to Decrypt");
+            Debug.Assert((offset >= 0) && (offset < buffer.Length), "Invalid input offset passed to Decrypt");
+            Debug.Assert((count > 0) && (count <= (buffer.Length - offset)), "Invalid input count passed to Decrypt");
+
+            using (SafeGssBufferHandle outputToken = new SafeGssBufferHandle())
+            {
+                libgssapi.Status minorStatus;
+                libgssapi.Status status = libgssapi.GssUnwrap(out minorStatus, context, buffer, offset, count, outputToken);
+                if (status != libgssapi.Status.GSS_S_COMPLETE)
+                {
+                    throw libgssapi.GssApiException.Create(SR.net_context_unwrap_failed, status, minorStatus);
+                }
+
+                int length = buffer.Length - offset;
+                if (outputToken.Length > length)
+                {
+                    throw libgssapi.GssApiException.Create(SR.Format(SR.net_context_buffer_too_small, outputToken.Length, length));
+                }
+
+                if (outputToken.Length > 0)
+                {
+                    Marshal.Copy(outputToken.Value, buffer, offset, outputToken.Length);
+                }
+
+                return outputToken.Length;
+            }
+        }
+
+        internal static string GetSourceName(SafeGssContextHandle context)
+        {
+            libgssapi.Status minorStatus;
+            SafeGssNameHandle sourceName;
+            libgssapi.Status status = libgssapi.GssInquireSourceName(
+                         out minorStatus,
+                         context,
+                         out sourceName);
+
+            if (status != libgssapi.Status.GSS_S_COMPLETE)
```

Comment:
This code is going to be removed. It is not supposed to be populated on error else gss_release_name will be needed. The sample code we have seen does not do this on errors